applications/main/nfc/plugins/supported_cards/ndef.c

```diff
@@ -86,29 +98,56 @@ static void parse_ndef_uri(FuriString* str, const uint8_t* payload, uint32_t pay
         prepend = prepends[prepend_type];
     }
     size_t prepend_len = strlen(prepend);
-
     size_t uri_len = prepend_len + (payload_len - 1);
-    char* const uri_buf = malloc(uri_len);
+
+    char* uri_buf = malloc(uri_len + 1);
+    if(!uri_buf) {
+        furi_string_cat_printf(str, "Memory allocation failed\n");
+        return;
+    }
+
     memcpy(uri_buf, prepend, prepend_len);
     memcpy(uri_buf + prepend_len, payload + 1, payload_len - 1);
-    char* uri = uri_buf;
+    uri_buf[uri_len] = '\0';
+
+    char* decoded_uri =
+        malloc(uri_len * 2 + 1); // Worst case scenario: every character is percent-encoded
```

Comment:
i think you got the right intent, but logic the wrong way around :D

if encoded its `"%20"`, 3 characters, decoded it becomes `" "`, so 1 character. so, worst case scenario, nothing was encoded, so length is the same. best case scenario, everything is encoded, length is roughly / 3. malloc() of same size should work fine?

Reply:
and because of this, i think it is possible to decode without a second buffer... iterate over same string keeping 2 indexes (original and decoded), for each iteration check char at original index: 
if percent: replace char at decoded index with decoded char, increase decoded index by 1, increase original index by 3
if any other character: read char at original index and write to decoded index
because the indexes will always either increase at same pace, or original increases faster than decoded, you can overwrite data because original index is ahead
not sure if im making sense lol

Reply:
but also depends if we want to show the original buffer...
fair, modern browsers will auto-encode if you type unsupported characters into the url bar, but still its not an accurate representation of what is on the card... eh should be fine

Reply:
> but also depends if we want to show the original buffer... fair, modern browsers will auto-encode if you type unsupported characters into the url bar, but still its not an accurate representation of what is on the card... eh should be fine

personally i don't see the need if it's functioning correctly and yeah that's a really good point avoiding a second buffer